design/hardware-status.md

```diff
@@ -0,0 +1,144 @@
+will regularly collect the data and post it to ironic-inspector, from where it
+can be consumed as part of the regular polling for status.
+
+## Motivation
+
+We need to be able to collect and update enough information about hardware
+devices in a server to monitor their status and identify them for
+maintenance and replacement.
+Valuable information are for example the current status, type, serial number,
+physical location, part number.
+We need to be able to refresh the information about the status of the hardware
+devices and, if a device has been replaced, we need to be able to retrieve the
+new information and update the hardware inventory to provide a constant
+up-to-date vision on the status of the devices.
+
+### Goals
+
+1. Decide which information on the devices to retrieve.
+2. Agree on how to retrieve such information.
+3. Provide a way to manually/automatically collect and update the devices
+information.
+
+### Non-Goals
+
+1. Any update on device data should not trigger any reaction from ironic.
+
+## Proposal
+
+### Implementation Details/Notes/Constraints
+
+- Depending on the device, we need to collect the following information, when
+applicable:
+    * Capacity
+    * Location
+    * Manufacturer
+    * Model
+    * Part Number
+    * Serial Number
+    * Status
+    * SMART details
+    
+- NVMe drives require special tools to get SMART data (e.g. nvme-cli),
+although [smartmontools](https://www.smartmontools.org/wiki/NVMe_Support)
+supports NVMe devices since version 6.5.
+ 
+- [Hdparm](https://en.wikipedia.org/wiki/Hdparm) can provide SMART data for
+most of drives. It doesn't work with NVMe devices.
+
+- [python-hardware](https://github.com/redhat-cip/hardware) can collect SMART
+data, although it requires code changes to detect and provide info on NVMe
+devices, for example relying directly on nvme tools instead of smartmontools.
+
+- [Ironic Python Agent (IPA)](https://github.com/openstack/ironic-python-agent)
+is already able to collect most of this information.
+Using a subset of the IPA code running inside a container in the hosts where we
+want to collect the drives data from can also allow to provide periodic
+updates on all the required info.
+The new containerized service will talk to Ironic Inspector.
+
+- [Ironic Inspector](https://docs.openstack.org/ironic-inspector/latest/)
+needs to be able to publish up-to-date data at any time, including from active
+and available nodes.
+This feature requires changes in the inspector code that currently accept data
+updates only from nodes in manageable state.
+
+### Risks and Mitigations
+
+1. Upstream Ironic community may reject changes in Ironic Inspector code.
+
+## Design Details
+
+### Work Items
+
+- Add option to detect and provide info on NVMe drives to python-hardware.
+- Extract and modify part of the IPA code to be able to run it in a container
+as a new service.
+- Modify Ironic Inspector code to accept data from active nodes.
```

Comment:
The baremetal operator uses gophercloud to get the inspection data. Will the operator need any of the new values being collected, or are all of those just going to go through the prometheus exporter as metrics? If the operator does need any of the new values, do we have to modify gophercloud to support them?

Reply:
@dhellmann we won't collect new values, we're going to extend support for a new type of device, nvme, but it won't change the data collected, the fields will be part of the disk info.

Reply:
@elfosardo do we already collect SMART data somewhere? I thought part of this was collecting runtime stats?

Reply:
@dhellmann The operator will likely need to identify when new disks appear as that information will need to be updated in the UI. Regarding smart, nvme is different and is just going to end up being yet another type of storage device.

Reply:
@juliakreger When you say "another type of storage device" I start to worry about things like whether the keys in the JSON payload coming from ironic-inspector are the same. Because if they're different, then we have to update gophercloud to look for those new keys explicitly in order to populate the data structure it defines and that the operator uses to identify something as a storage device.

Reply:
@dhellmann That is a very valid concern because it is true, although I thought we left the high level structures for introspection data such that it was not on the low level keys.

Ultimately, @elfosardo needs time with python-hardware and an nvme storage device to provide a clearer picture.

Reply:
OK. It would be best if we could use the same names for those keys, unless they're just "wrong" somehow.

Reply:
@dhellmann currently python-hardware exports information on smart data using the `SMART` keyword. This is parsed by ironic-python-agent, jsonized and transmitted to ironic inspector.
The keys all have the `SMART` prefix, for example (taken from an actual device):

{
  "sda{megaraid,0}": {
    "SMART/product": "ST600MM0006",
    "SMART/start_stop_cycle_count": 739,
    "SMART/current_drive_temperature": 27,
    "SMART/blocks_sent": 2456935760,
    [.....]
    "SMART/health": "OK",
    "SMART/blocks_received": 2118230176,
    "SMART/non_medium_errors_count": 31,
    "SMART/specified_start_stop_cycle_count_over_lifetime": 10000,
 
This behaviour won't change for nvme devices.

Reply:
Sounds good.